scene/main/viewport.cpp

```diff
@@ -859,16 +859,38 @@ bool Viewport::_is_size_allocated() const {
 Rect2 Viewport::get_visible_rect() const {
 	Rect2 r;
 
-	if (size == Size2()) {
-		r = Rect2(Point2(), DisplayServer::get_singleton()->window_get_size());
-	} else {
+	bool subViewport_exist = false;
+	Node *temp = get_node(get_path())->get_child(0, true);
+
+	while (temp != NULL) {
+		if (get_tree()->get_edited_scene_root() != NULL && temp != NULL) {
+			if (get_tree()->get_edited_scene_root() == temp) {
+				break;
+			}
+		}
+
+		if (temp->get_class() == "SubViewport") {
+			subViewport_exist = true;
+			break;
+		}
+		else {
+			subViewport_exist = false;
+		}
+
+		temp = temp->get_parent();
+	}
+
+	if (subViewport_exist == true) {
 		r = Rect2(Point2(), size);
```

Comment:
Finally, I checked what is the easiest way to check if the camera belongs to the editor's SubViewport, and this is what I got instead of all these changes:
```diff
diff --git a/scene/main/viewport.cpp b/scene/main/viewport.cpp
index 1287f69bef..992cfd65e5 100644
--- a/scene/main/viewport.cpp
+++ b/scene/main/viewport.cpp
@@ -864,8 +864,8 @@ bool Viewport::_is_size_allocated() const {
 Rect2 Viewport::get_visible_rect() const {
 	Rect2 r;
 
-	if (size == Size2()) {
-		r = Rect2(Point2(), DisplayServer::get_singleton()->window_get_size());
+	if (get_tree()->get_edited_scene_root() && get_tree()->get_edited_scene_root()->get_parent() == get_viewport()) {
+		r = Rect2(Point2(), Size2(GLOBAL_GET("display/window/size/viewport_width"), GLOBAL_GET("display/window/size/viewport_height")));
 	} else {
 		r = Rect2(Point2(), size);
 	}

```

I think it will be better than the while loop. But perhaps you should check it yourself a few more times with an example [Camera3D Frustum.zip](https://github.com/godotengine/godot/files/10366473/Camera3D.Frustum.zip)

https://user-images.githubusercontent.com/7782218/211155734-bf1c86a2-7853-4588-8d35-08753ac340da.mp4

Reply:
yeah this seems great ! 
Sorry you did all the work 😅 But i learned several things. thanks.